src/Pos/Wallet/Web/ClientTypes.hs

```diff
@@ -135,15 +140,16 @@ convertTxOutputs :: [TxOut] -> [(CAddress, CCoin)]
 convertTxOutputs = map (addressToCAddress . txOutAddress &&& mkCCoin . txOutValue)
 
 mkCTx
-    :: Address            -- ^ An address for which transaction info is forming
+    :: (MonadThrow m, MonadTxExtraRead m)
+    => Address            -- ^ An address for which transaction info is forming
     -> ChainDifficulty    -- ^ Current chain difficulty (to get confirmations)
     -> TxHistoryEntry     -- ^ Tx history entry
     -> CTxMeta            -- ^ Transaction metadata
-    -> CTx
-mkCTx addr diff THEntry {..} meta =
-    -- placeholder, this functionality isn't implemented yet
-    let ctFrom = [] in
-    CTx {..}
+    -> m CTx
+mkCTx addr diff THEntry {..} meta = do
+    txExtra <- maybeThrow (Internal "Transaction is not in the DB") =<< getTxExtra _thTxId
+    let ctFrom = convertTxOutputs $ map toaOut $ toList $ teInputOutputs txExtra
+    pure CTx {..}
```

Comment:
Well, it's probably my fault, I should have thought about the actual task before consulting you on how to inject database access into `WalletWebMode`.
It would be way more clean to extend `TxHistoryEntry` to contain input/output addresses as well, because while deriving history we have access to all the necessary info about them. And it seems like it shouldn't be hard...
It's mostly my fault, it was way harder for you to come up with this solution, because you're unfamiliar with related code. We can probably leave this code as is, but it woud be much better to update functions which derive history instead.